File: Data Structures & Algorithms/balanced-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int dfs(TreeNode root) {
        if (root == null) return 0;
        int left = dfs(root.left);
        int right = dfs(root.right);
        return 1 + Math.max(left, right);
    }

    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        if (Math.abs(dfs(root.left) - dfs(root.right)) > 1) return false;
        return isBalanced(root.left) && isBalanced(root.right);
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(node, leftMin, rightMax):
            if not node:
                return True

            if not (leftMin < node.val < rightMax):
                return False

            return isValid(node.left, leftMin, node.val) and isValid(node.right, node.val, rightMax)

        return isValid(root, float("-inf"), float("inf"))